### ISSUE TITLE
[v2] Ensure components don't access `ref` for React 19 compatibility

### ISSUE THREAD

**Issue by Philzen** [chakra-ui/chakra-ui#9329]
### Description

In react 19 any components ~~could be conditionally exported without being wrapped in `forwardRef` when React v19 is used.~~ should refrain from `element.ref` access.

### Problem Statement/Justification

In React 19, any components ~~using `forwardRef`~~ [accessing `element.ref` will trigger a warning](https://react.dev/blog/2024/04/25/react-19-upgrade-guide#deprecated-element-ref):

> Accessing element.ref was removed in React 19. ref is now a regular prop. It will be removed from the JSX Element type in a future release.

[This is because `ref` has become a regular prop in React 19](https://react.dev/blog/2024/12/05/react-19#ref-as-a-prop). 

Side note: This is actually nice b/c it removes the need to wrap any components in `forwardRef` (something noteworthy to add to the docs).

<details><summary>forwardRef migration for libs that want it</summary>
<p>

```tsx
import { version } from 'react'

function _Tooltip (props) {
  // …
  {
    // …
    ref,
    ...rest
  } = ownProps
  // …
}

export const Tooltip = version.startsWith('19.') 
  ? _Tooltip 
  : forwardRef<TooltipProps, "div">((props, ref) => _Tooltip({ ...props, ref }))
```

</p>
</details> 

### Proposed Solution or API

A commonly used component that triggers the warning is the popular chakra `<Tooltip />`:

https://github.com/chakra-ui/chakra-ui/blob/ebf39c93d1a88982d14230d69dd9404f18b4f8b9/packages/components/src/tooltip/tooltip.tsx#L125-L128

The offending bit is `child.ref` which used to be a normal field, but in React 19 was changed to a getter that will still return the value, but trigger the warning:

```
Object.getOwnPropertyDescriptor(temp0, "ref").get // function elementRefGetterWithDeprecationWarning()
```

Not relying on this getter being there forever, nor it's name, the most stable future-proof solution i could come up with is this one-line change

```diff
      tooltip.getTriggerProps(
        child.props,
-       child.ref,
+       Object.getOwnPropertyDescriptor(child, 'ref').value
      )
```

`Object.getOwnPropertyDescriptor(child, 'ref').value` will return the `child.ref` in react versions before 19, and `undefined` starting with react 19 – which is what we want b/c there it is now available as `child.props.ref`.

### Alternatives

Above proposed solution basically works like a charm in initial tests. However i'm not aware if there may be any drawbacks here ~~(i.e. will treeshaking still work the same?)~~.

Alternatives i could think of that all work without evaluating `react.version` or the name of the getter function in the code itself:
1. Provide a separate package  
If there is some kind of npm mechanism to resolve to different package versions depending on what react peerDependency version is detected, that would be awesome to automatically provide the correct one. Otherwise, not a big deal to have developers manually change package.json to use the react19-compatible chakra package when they migrating.  
Of course there is the drawback of maintaining this.
2. Provide react19-compatible component versions as separate imports  
~~So alongside of the existing Tooltip export we could introduce a second one not wrapped in `forwardRef` called `Tooltip19` for instance. Would definitely work very clean with treeshaking.~~
3. ~~Provide a wrapper component that has a prop for this  
So if that prop (i.e. `noForwardRef`) on the wrapper is set, the child will be rendered accordingly. Again, not sure how well that works with treeshaking.~~  
Evaluate the react version in code  
    ```tsx  
    import { version } from 'react'
  
    //…
    
          tooltip.getTriggerProps(
            child.props,
            version.startsWith('19.') ? undefined : child.ref
          )
    ```

**Comment by Philzen** [comment]
@segunadebayo if the chakra team approves … depending on which implementation you favor – i'd be happy to contribute a PR

**Comment by Philzen** [comment]
After some more testing, i realized i was barking up the wrong tree :wink:

The culprit is not `forwardRef` (which is planned for deprecation in future version but currently does not issue any warnings), but accessing `element.ref`. I have now updated the issue title and description accordingly.

**Comment by segunadebayo** [comment]
@Philzen We're definitely interested in fixing this. Will have a look this week

**Comment by Philzen** [comment]
> @Philzen We're definitely interested in fixing this. Will have a look this week

I just realized i hadn't updated the issue title accordingly as intended. Conditionally removing `forwardRef`-wrapping may be a nice optimization (and also a measure to stay future-proof). If you're interested, i can update a separate PR for that.

To fix the issue at hand i have whipped up https://github.com/chakra-ui/chakra-ui/pull/9333, which has already been approved by @jcpendesign but is missing an unrelated CI check authorization. 

Once that is merged i'm happy to forward-port that fix to Chakra v3 if you like.

**Comment by segunadebayo** [comment]
I already fixed the v3 since you brought it up. Thanks again.

For v2, it's now done. See https://github.com/chakra-ui/chakra-ui/pull/9333#issuecomment-2598891188

Will do a new release shortly.
